### DISCUSSION TITLE
STM32L uart RX not working with echo sample

### DISCUSSION THREAD

**Discussion by jasshapiro** [zephyrproject-rtos/zephyr#69639] [Q&A] [answered]
Hello zephyr community,

I'm converting an application from a stm32f405 express to stm32L4+ swan from blues. Both are supported in tree boards.
While attempting to move the zephyr console from CDC to hardware uarts {lpuart1, usart1}, I found that my laptop could receive characters from the stm32L uController, but I could net send traffic in the other direction

To factor out my code, I built samples/subsys/console/echo and got the same result. Boot prompt from the application, but no echo. Scoping this uC's RX line, I can see a the pulse train from the laptop, but amplitude of the pulse is only about 0.3V. As though the uC is pulling the TTL line up to hard.

I swapped out USB/Serial converters and uControllers to confirm that it's not an issue in the test setup. 

I came across [this old issue](https://github.com/zephyrproject-rtos/zephyr/issues/39395 ). And it seems to have closed with no resolution. As suggested there, usart3 works as expected, usart2 also works as expected. While USART1, and LPUART1 only transmit and do not receive.  I have tried to do things like bias rx pin (which seems to be pulled up hard) down, but this had no effect.

Any help or direction for troubleshooting would be appreciated.

Thanks,
-Justin

**Comment by jasshapiro** [comment]
Discovered that LPUART1 is connected to the JTAG connected. Confirmed that it is working when jtag disconnected. 
USART1 still has no RX

**Reply by ndrs-pst** [reply]
@jasshapiro Looking at `zephyr.dts` might be the first step I take. Since the `uart_stm32.c` code is shared among the STM32 series, the chances are low that the issue is caused by the driver itself. However, if you can share `zephyr.dts`, it would be helpful.

**Reply by jasshapiro** [reply]
Hi, Sorry for delay as priorities shift back to here.
See attached zephyr.dts (with .txt for github).
usart1 is on pa9/pa10 which agrees with the schematic.

<img width="345" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/20733929/503e3035-3963-4fdd-b21a-e8b9190a7c5a">

```
        usart1: feather_serial: serial@40013800 {                                  
            compatible = "st,stm32-usart", "st,stm32-uart";                        
            reg = < 0x40013800 0x400 >;                                            
            clocks = < &rcc 0x60 0x4000 >;                                         
            resets = < &rctl 0x80e >;                                              
            interrupts = < 0x25 0x0 >;                                             
            status = "okay";                                                       
            pinctrl-0 = < &usart1_tx_pa9 &usart1_rx_pa10 >;                        
            pinctrl-names = "default";                                             
            current-speed = < 0x1c200 >;                                           
        }; 
```
[zephyr.dts.txt](https://github.com/user-attachments/files/15916873/zephyr.dts.txt)

Thanks,
-J

**Reply by ndrs-pst** [reply, marked as answer]
One thing I suspect out is the following with `usb_otg_fs_id_pa10` and status = "okay";
Even it shouldn't be the problem if you didn't enable the USB driver !!!

```
usbotg_fs: zephyr_udc0: otgfs@50000000 {
	compatible = "st,stm32-otgfs";
	reg = < 0x50000000 0x40000 >;
	interrupts = < 0x43 0x0 >;
	interrupt-names = "otgfs";
	num-bidir-endpoints = < 0x6 >;
	ram-size = < 0x500 >;
	maximum-speed = "full-speed";
	clocks = < &rcc 0x4c 0x1000 >, < &rcc 0x5 0x7a88 >;
	phys = < &otgfs_phy >;
	status = "okay";
	pinctrl-0 = < &usb_otg_fs_dm_pa11 &usb_otg_fs_dp_pa12 &usb_otg_fs_id_pa10 >;
	pinctrl-names = "default";
};
```

Also, finally you will change 

```
zephyr,console = &lpuart1;
zephyr,shell-uart = &lpuart1;
```

to &usart1 right ?

**Reply by jasshapiro** [reply]
Hi,

I see this conflict, and i think it comes up when also enable the CDC uart. If I fully remove USB/CDC from the device tree and prj.conf, pa10 is freed and available for by RX, so problem solved.

For my application, I'd ideally use both USB_CDC and UART0-3. I cannot se another pin that  `usb_otg_fs_id_pa10` can be muxed to. Do you know of another way to enable these functions simultaneously?

Thanks.

**Reply by ndrs-pst** [reply]
Just change the device tree in the `.overlay` file like this (remove `&usb_otg_fs_id_pa10`), and you should be good to go. 
Since for now Zephyr primarily supports just USB devices and is currently working to enhance this: https://github.com/zephyrproject-rtos/zephyr/issues/42066

```
usbotg_fs: zephyr_udc0: otgfs@50000000 {
	pinctrl-0 = < &usb_otg_fs_dm_pa11 &usb_otg_fs_dp_pa12 >;
	pinctrl-names = "default";
};
```

**Reply by jasshapiro** [reply]
Thank you. Adding in the overlay
```
&usbotg_fs{                                                                        
    pinctrl-0 = < &usb_otg_fs_dm_pa11 &usb_otg_fs_dp_pa12>;                        
    status="okay";                                                                 
}; 
```
Gets the USB CDC pseudo uart running again along with USART1. In summary, the issue was enabling CDC and USART1 without de-obligating  `usb_otg_fs_id_pa10` from the `usbotg_fs` node. 

Separate question, I had the incorrect assumption that pinctrl-N phandles were position and required as they can correspond with hardware/signal lines, I had also thought this was a way for the build system to enforce SOC configuration compatibility with kernel drivers. Is this typically handled a level up, at the kernel driver level?

Thank you for your help. I'm actually quite surprised to have worked through this issue.

**Reply by ndrs-pst** [reply]
Yes, normally the driver uses this information to set the pin alternative via `pinctrl_apply_state`, while the use of the `index-N`, as far as I have seen, is to apply pin state according to power management.

https://github.com/zephyrproject-rtos/zephyr/blob/72197097aeb75ec9a7efe716d8822b0a72e7cfc5/boards/st/nucleo_wba55cg/nucleo_wba55cg.dts#L115-L123

https://github.com/zephyrproject-rtos/zephyr/blob/72197097aeb75ec9a7efe716d8822b0a72e7cfc5/include/zephyr/devicetree/pinctrl.h#L84-L105
